src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -3480,5 +3486,425 @@ void F()
+    {
+        Action a1 = [DllImport(""MyModule.dll"")] static () => { };
+        Action a2 = [DynamicSecurityMethod] () => { };
+        Action a3 = [SuppressUnmanagedCodeSecurity] () => { };
+        Func<object> a4 = [return: MarshalAs((short)0)] () => null;
+    }
+}";
+            var sourceB =
+@"namespace System.Security
+{
+    internal class DynamicSecurityMethodAttribute : Attribute { }
+}";
+            var comp = CreateCompilation(new[] { sourceA, sourceB }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (8,22): error CS0601: The DllImport attribute must be specified on a method marked 'static' and 'extern'
+                //         Action a1 = [DllImport("MyModule.dll")] static () => { };
+                Diagnostic(ErrorCode.ERR_DllImportOnInvalidMethod, "DllImport").WithLocation(8, 22));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var exprs = tree.GetRoot().DescendantNodes().OfType<LambdaExpressionSyntax>().ToImmutableArray();
+            Assert.Equal(4, exprs.Length);
+            var lambdas = exprs.SelectAsArray(e => GetLambdaSymbol(model, e));
+            Assert.Null(lambdas[0].GetDllImportData()); // [DllImport] is ignored if there are errors.
+            Assert.True(lambdas[1].RequiresSecurityObject);
+            Assert.True(lambdas[2].HasDeclarativeSecurity);
+            Assert.Equal(default, lambdas[3].ReturnValueMarshallingInformation.UnmanagedType);
+        }
+
+        [Fact]
+        public void LambdaAttributes_Permissions()
+        {
+            var source =
+@"#pragma warning disable 618
+using System;
+using System.Security.Permissions;
+class Program
+{
+    static void Main()
+    {
+        Action a1 = [PermissionSet(SecurityAction.Deny)] () => { };
+    }
+}";
+            var comp = CreateCompilationWithMscorlib40(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var exprs = tree.GetRoot().DescendantNodes().OfType<LambdaExpressionSyntax>().ToImmutableArray();
+            var lambda = exprs.SelectAsArray(e => GetLambdaSymbol(model, e)).Single();
+            Assert.NotEmpty(lambda.GetSecurityInformation());
+        }
+
+        [Fact]
+        public void LambdaAttributes_NullableAttributes()
+        {
+            var source =
+@"using System;
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static void Main()
+    {
+        Func<object> a1 = [return: MaybeNull][return: NotNull] () => null;
+        Func<object, object> a2 = [return: NotNullIfNotNull(""obj"")] (object obj) => obj;
+        Action a3 = [DoesNotReturn] () => { };
+        Func<bool> a4 = [MemberNotNull(""x"")][MemberNotNullWhen(false, ""y"")][MemberNotNullWhen(true, ""z"")] () => true;
+    }
+}";
+            var comp = CreateCompilation(
```

Comment:
Prototype comment to make sure these actually do the things they're supposed to?